### PR TITLE
Fixing some questionable uses of Type::subst()

### PR DESCRIPTION
**Pull request by slavapestov**

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

File: lib/SILOptimizer/Utils/Generics.cpp [hunk 1]
```diff
@@ -644,62 +644,13 @@ checkSpecializationRequirements(ArrayRef<Requirement> Requirements) {
...
-        return SpecializedGenericEnv->mapTypeOutOfContext(
-          SpecializedGenericEnv->mapTypeIntoContext(type));
-      },
-      LookUpConformanceInSignature(*SpecializedGenericSig));
-  }
-
-  OriginalParamSubs = CallerParamSubs;
-
-  HasUnboundGenericParams = !SpecializedGenericSig->areAllParamsConcrete();
-  createSubstitutedAndSpecializedTypes();
```
Comment (swiftix) [inline lib/SILOptimizer/Utils/Generics.cpp]: Removing all this code from the constructor will break explicit specialization which is used by @_specialize (this constructor is invoked by the EagerSpecializer).

Comment (slavapestov) [inline lib/SILOptimizer/Utils/Generics.cpp]: That should be refactored to use the new methods you added though right?

Comment (swiftix) [inline lib/SILOptimizer/Utils/Generics.cpp]: Probably, but explicit and automatic partial specializations have rather different setups. So, please leave this code there for now. I'll re-factor myself once we are there.

Comment (slavapestov) [inline lib/SILOptimizer/Utils/Generics.cpp]: Ok, I'll put it back and add a FIXME.

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
https://github.com/apple/swift-corelibs-foundation/pull/951

@swift-ci Please smoke test Linux

**Comment by slavapestov** [comment]
@swift-ci Please smoke test Linux

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Pull request merged**